File: Python/Problems/Problems/problem2786.py
import heapq
import math
import sys

from utils import print1DArray, print2DArray
from typing import List, Set, Optional
from collections import Counter

from .problem import Problem


class Problem2786(Problem):
    def test(self) -> bool:
        nums = [2, 3, 6, 1, 9, 2]
        x = 5

        expected = 13

        result = self.maxScore(nums, x)

        print(result)

        return result == expected

    def maxScore(self, nums: List[int], x: int) -> int:
        n = len(nums)
        even = nums[0] if nums[0] % 2 == 0 else -x
        odd = nums[0] if  nums[0] % 2 == 1 else -x
        for i in range(1, n):
            if nums[i] % 2 == 0:
                even = max(odd + nums[i] - x, even + nums[i])
            else:
                odd = max(even + nums[i] - x, odd + nums[i])

        return max(even, odd)

File: C++/Problems/Problems/problem_2786.cpp
#include "problem_2786.h"

bool problem_2786::test() {
    std::vector nums { 2, 3, 6, 1, 9, 2 };
    int x = 5;

    int expected = 13;

    auto result = maxScore(nums, x);

    std::cout << result << std::endl;

    return result == expected;
}

long long problem_2786::maxScore(std::vector<int> &nums, int x) {
    int n = nums.size();
    long even = nums[0] % 2 == 0 ? nums[0] : -x;
    long odd = nums[0] % 2 == 1 ? nums[0] : -x;
    for (int i = 1; i < n; ++i)
    {
        if (nums[i] % 2 == 0)
            even = std::max(odd + nums[i] - x, even + nums[i]);
        else
            odd = std::max(even + nums[i] - x, odd + nums[i]);
    }

    return std::max(even, odd);
}
